### PR TITLE
DNP3 Support - v7

### PR DESCRIPTION
**Pull request by jasonish**
Replaces #1791.

This PR contains:
An application decoder for DNP3.
* Snort compatible DNP3 keywords:
  * dnp3_data
  * dnp3_func
  * dnp3_ind
  * dnp3_obj
* A DNP3 output module for EVE (JSON logging).
* DNP3 support for Lua detection.

Documentation for the DNP3 rule keywords can be found here:
https://redmine.openinfosecfoundation.org/projects/suricata/wiki/DNP3_keyword
OR in newer format:
http://jasonish-suricata.readthedocs.org/en/latest/rules/dnp3-keywords.html

Changes from the last PR include:
* Misc cleanups.
* More objects.
* Generating object logging code.

PRScript output:
- PR jasonish-pcap: https://buildbot.openinfosecfoundation.org/builders/jasonish-pcap/builds/184
- PR jasonish: https://buildbot.openinfosecfoundation.org/builders/jasonish/builds/186

### CODE REVIEW

File: rules/dnp3-events.rules [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Flooded.
```
Comment (inliniac) [inline rules/dnp3-events.rules]: File misses a header like http-events.rules has

File: src/util-error.h [hunk 2]
```diff
@@ -271,6 +271,7 @@ typedef enum {
     SC_ERR_THRESHOLD_SETUP,
     SC_ERR_DNS_CONFIG,
     SC_ERR_MODBUS_CONFIG,
+    SC_ERR_DNP3_CONFIG,
```
Comment (inliniac) [inline src/util-error.h]: new errors should be appended to keep the error codes stable

File: src/output-json-dnp3.c [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+    json_t *jsitems;
+
+    if (unlikely((jsitems = json_array()) == NULL)) {
+        return NULL;
+    }
+
+    TAILQ_FOREACH(item, object->points, next) {
+        json_t *js = json_object();
+        if (unlikely(js == NULL)) {
+            break;
```
Comment (inliniac) [inline src/output-json-dnp3.c]: jsitems is lost here

File: src/output-json-dnp3.c [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+#include "detect-dnp3.h"
+
+#include "output.h"
+#include "output-json.h"
+#include "output-json-dnp3-objects.h"
+
+#ifdef HAVE_LIBJANSSON
+#include <jansson.h>
+
+/* Appears not all current distros have jansson that defines this. */
```
Comment (inliniac) [inline src/output-json-dnp3.c]: move this to suricata-common.h?

File: src/output-json-dnp3.c (hunk 3 above)
Comment (jasonish) [inline src/output-json-dnp3.c]: Its still returned even in this case, which means it gets referenced by the parent object, even if empty, so will get free'd with the complete object.

Perhaps that would be more clear if we allocated the json_t object in the caller and passed it in?

Comment (inliniac) [inline src/output-json-dnp3.c]: Looks like I misread the code.

**Comment by inliniac** [comment]
This will need AFL entry points similar to https://github.com/inliniac/suricata/pull/1999/files#diff-f5e72d82198f985531c162cd7fca0b54R1430

Then we'll hammer it with some fuzz testing.

**Pull request closed without merging**